aquadoggo/src/libp2p/config.rs

```diff
@@ -0,0 +1,163 @@
+
+/// Libp2p config for the node.
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct Libp2pConfiguration {
+    /// Dial concurrency factor.
+    ///
+    /// Number of addresses concurrently dialed for a single outbound
+    /// connection attempt.
+    pub dial_concurrency_factor: u8,
+
+    /// Local address.
+    pub listening_multiaddr: Multiaddr,
+
+    /// Maximum incoming connections.
+    pub max_connections_in: u32,
+
+    /// Maximum outgoing connections.
+    pub max_connections_out: u32,
+
+    /// Maximum pending incoming connections.
+    pub max_connections_pending_in: u32,
+
+    /// Maximum pending outgoing connections.
+    pub max_connections_pending_out: u32,
+
+    /// Maximum connections per peer (includes outgoing and incoming).
+    pub max_connections_per_peer: u32,
+
+    /// mDNS discovery enabled.
+    pub mdns: bool,
+
+    /// Notify handler buffer size.
+    ///
+    /// Defines the buffer size for events sent from the NetworkBehaviour to the ConnectionHandler.
+    /// If the buffer is exceeded, the Swarm will have to wait. An individual buffer with this
+    /// number of events exists for each individual connection.
+    pub notify_handler_buffer_size: usize,
+
+    /// Connection event buffer size.
+    ///
+    /// Defines the buffer size for events sent from the ConnectionHandler to the
+    /// NetworkBehaviour. Each connection has its own buffer. If the buffer is
+    /// exceeded, the ConnectionHandler will sleep.
+    pub per_connection_event_buffer_size: usize,
+
+    /// Ping behaviour enabled.
+    pub ping: bool,
+}
+
+impl Default for Libp2pConfiguration {
+    fn default() -> Self {
+        // Define the default listening multiaddress using the default QUIC port
+        let listening_multiaddr = format!("/ip4/0.0.0.0/udp/{QUIC_PORT}/quic-v1")
+            .parse()
+            .unwrap();
+
+        Self {
+            dial_concurrency_factor: 8,
+            listening_multiaddr,
+            max_connections_in: 16,
+            max_connections_out: 16,
+            max_connections_pending_in: 8,
+            max_connections_pending_out: 8,
+            max_connections_per_peer: 8,
+            mdns: false,
+            notify_handler_buffer_size: 128,
+            per_connection_event_buffer_size: 8,
+            ping: false,
+        }
+    }
+}
+
+impl Libp2pConfiguration {
+    /// Define the connection limits of the swarm.
+    pub fn connection_limits(&self) -> ConnectionLimits {
+        ConnectionLimits::default()
+            .with_max_pending_outgoing(Some(self.max_connections_pending_out))
+            .with_max_pending_incoming(Some(self.max_connections_pending_in))
+            .with_max_established_outgoing(Some(self.max_connections_out))
+            .with_max_established_incoming(Some(self.max_connections_in))
+            .with_max_established_per_peer(Some(self.max_connections_per_peer))
+    }
+
+    /// Load the keypair from the file at the specified path.
```

Comment:
Extra newline here, then it will render nicely in RustDoc previews :sparkles:

Reply:
Super nitpick: We spell "key pair" or `key_pair` separated everywhere else. Something I notice because we talked about it for ages in the past haha